### PR TITLE
Using gradle ExternalDependencyModule changing attribute to fix #694

### LINKED ISSUE #694: Use Gradle `changing` to determine snapshots
From 
https://github.com/GoogleContainerTools/jib/pull/584#discussion_r204366857.  Gradle supports [explicitly indicating that a dependency may be changing](
https://stackoverflow.com/a/42058780/600339) and [provides an API](https://docs.gradle.org/current/javadoc/org/gradle/api/artifacts/ExternalModuleDependency.html#isChanging--) to check.  The [docs](https://docs.gradle.org/current/userguide/declaring_dependencies.html#sub:declaring_dependency_with_changing_version) [hint](https://docs.gradle.org/current/userguide/troubleshooting_dependency_resolution.html#sub:dynamic_versions_and_changing_modules) that SNAPSHOTs are marked as changing.

### PR DESCRIPTION
**Pull request by rajovictor**
Fixes #694

### CODE REVIEW

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleProjectProperties.java [hunk 1]
```diff
@@ -173,6 +174,22 @@ public JibContainerBuilder createJibContainerBuilder(
 
       FileCollection snapshotDependencies =
           nonProjectDependencies.filter(file -> file.getName().contains("SNAPSHOT"));
+
+      FileCollection changingDependencies =
+          project.files(
+              project
+                  .getConfigurations()
+                  .getByName(JavaPlugin.RUNTIME_CLASSPATH_CONFIGURATION_NAME)
+                  .getDependencies()
```
Comment (chanseokoh) [inline jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleProjectProperties.java]: I see these lines are duplicated above, so I'd like to see this part factored out as a variable. Like

```java
      Configuration runtimeClasspathConfiguration =
          project.getConfigurations().getByName(JavaPlugin.RUNTIME_CLASSPATH_CONFIGURATION_NAME);

      FileCollection projectDependencies =
          project.files(
              runtimeClasspathConfiguration
                  .getResolvedConfiguration()
      ...

      FileCollection changingDependencies =
          project.files(
              runtimeClasspathConfiguration
                  .getDependencies()
```

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleProjectProperties.java [hunk 2]
```diff
@@ -173,6 +174,22 @@ public JibContainerBuilder createJibContainerBuilder(
...
+                  .stream()
+                  .filter(artifact -> artifact instanceof ExternalModuleDependency)
+                  .filter(
+                      moduleDependency ->
+                          ((ExternalModuleDependency) moduleDependency).isChanging())
+                  .collect(Collectors.toList()));
+
+      snapshotDependencies.plus(changingDependencies);
```
Comment (chanseokoh) [inline jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleProjectProperties.java]: Based on [this](https://github.com/GoogleContainerTools/jib/issues/694#issue-343698282), I feel like this will end up having duplicate entries if the filename contains `SNAPSHOT`. I'd say first we need to verify whether it is really true that a file containing `SNAPSHOT` sets the "changing" mark. And if so, probably we don't need to do `nonProjectDependencies.filter(file -> file.getName().contains("SNAPSHOT")` anymore and just put `isChanging` into the snapshot layer directly.

That said, this kind of uncertainty is one of many reasons we need to add tests. Could you look at `GradleProjectPropertiesTest` and think of some tests? We can't really merge code that is untested.

Comment (rajovictor) [reply jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleProjectProperties.java]: I installed `jib-gradle-plugin:1.7.1-SNAPSHOT` to local maven and used in a sample project to verify that `SNAPSHOT` dependencies does not have `isChanging`.
I also added a test case to build the tar, extract layers and verify that both the `SNAPSHOT` jar and changing jar are in the same layer.

I also tried to build the project like below instead of mocking.
```
 Project layersProject = ProjectBuilder.builder().withProjectDir(new 
 File("test/resources/gradle/proejects/gradle-layers")).withName("root").build();
 GradleProjectProperties gradleProjectProperties = new GradleProjectProperties(project, 
 mockLogger);
    JavaContainerBuilder javaContainerBuilder =
            JavaContainerBuilder.from(RegistryImage.named("base"))
                    .setAppRoot(AbsoluteUnixPath.get("/app"))
                    .setModificationTimeProvider((ignored1, ignored2) -> 
                      SAMPLE_FILE_MODIFICATION_TIME);
   JibContainerBuilder jibContainerBuilder =  
gradleProjectProperties.createJibContainerBuilder(javaContainerBuilder,
DEFAULT_CONTAINERIZING_MODE);
 BuildConfiguration configuration =
            JibContainerBuilderTestHelper.toBuildConfiguration(
                    jibContainerBuilder,
                    Containerizer.to(RegistryImage.named("to"))
                            .setExecutorService(MoreExecutors.newDirectExecutorService()));
    ContainerBuilderLayers layers = new ContainerBuilderLayers(configuration);
```
But the project object was not completely built so didn't give the expected result.

Main problem I noticed while trying to write the unit tests is, the current `FileCollections` are all mocked but the FileCollection I am trying to build has to come from `DefaultExternalModuleDependency` object and I could not find a way to mock all the involved Gradle objects.

Bottom line, I could verify the functionality using external project and a tar build junit, but with those changes `com.google.cloud.tools.jib.gradle.GradleProjectPropertiesTest#testCreateContainerBuilder_correctFiles` test case is failing.
I need help.

Comment (chanseokoh) [reply jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleProjectProperties.java]: @rajovictor thanks for looking into the matter. About writing a unit test, I'll check out your test code in your comment and think of something, but it may take some time.

Comment (rajovictor) [reply jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleProjectProperties.java]: BTW, I also had to make few changes to the way I was checking and building the changingDependencies collection but will commit once we get the junits working.

Comment (chanseokoh) [reply jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleProjectProperties.java]: I've looked into this, and the simplest way will be to add an integration test and forget about adding some in `GradleProjectProperties`.

`GradleLayerConfigurationIntegrationTest` builds a sophisticated sample project at `jib-gradle-plugin/src/test/resources/gradle/projects/all-local-multi-service` and verifies each file in the image layers. I think you could define some changing dependencies in `jib-gradle-plugin/src/test/resources/gradle/projects/all-local-multi-service/complex-service/build.gradle` and verify if they go into the SNAPSHOT layer.

To run integration tests, you will do `./gradlew integrationTest`. But for it to work, the `JIB_INTEGRATION_TESTING_LOCATION` environment variable should be defined first. See `CONTRIBUTING.md` for the details.

Comment (rajovictor) [reply jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleProjectProperties.java]: Thank you for the reply.
After some attempts, I got the full tar and the container build (till layer construction) executed with successful dependency and snapshot verification.
But then due to the changes in `GradleProjectProperties` current Unit tests were failing with NPEs  on some of `filecollections`.
Lemme spend some more time on them to see if I can get them to working.

**Comment by loosebazooka** [comment]
I'm not sure exactly what was decided here. But by default we should include anything named "SNAPSHOT" in snapshot dependency layers, no?

**Comment by chanseokoh** [comment]
yeah, we still have `nonProjectDependencies.filter(file -> file.getName().contains("SNAPSHOT"))`, so they will continue to exist in the snapshot dep layer. And @rajovictor figured out those JARs are not marked with `isChanging`. Both SNAPSHOT plus `isChanging` will go into the snapshot dep layer.

**Comment by chanseokoh** [comment]
@rajovictor could you rebase the branch? And there are conflicts to resolve too.

**Comment by chanseokoh** [comment]
Actually, some changes are in compete with #2107, so perhaps it's better to wait until #2107 is merged.

**Comment by chanseokoh** [comment]
#2107 is merged. Could you merge with master and resolve conflicts?

**Comment by rajovictor** [comment]
My fork and master have diverged so much, I think, I will close this PR and create a fresh one, if that's ok.

**Comment by chanseokoh** [comment]
Absolutely! Just create a new one.

**Comment by rajovictor** [comment]
Closing the PR since there are too many merge conflicts.
